From 02c0539a86262982ff30b469d5e3c68371d3a0ff Mon Sep 17 00:00:00 2001
From: Sami Shalayel <sami.shalayel@qt.io>
Date: Wed, 1 Oct 2025 13:39:31 +0200
Subject: [PATCH 284/553] qdatetime: reject 2 digit years as invalid in rfc2822
 format

Reject all digit year notation in rfcDateImpl that don't use 4 digits,
and return an invalid date.
Using 2 digit year is marked as obsolete in RFC2822, and the previous
implementation interpreted 2 digit years as is, like 22 became 22 (not
1922, not 2022, but 22) for example.

Also note that the documentation already explicitly says that the year
should be in the form of "yyyy" and therefore contain 4 digits.

This helps fix a bug in qtdeclarative's date parser in
qv4dateobject.cpp, where we only apply our special fallbacks (for
example for 2 digit years) when the QDateTime parsing methods return
an invalid QDateTime.

Pick-to: 6.8
Task-number: QTBUG-100377
Change-Id: I12c61ca35f38797e954e4126763252026ecbb008
Reviewed-by: Edward Welbourne <edward.welbourne@qt.io>
(cherry picked from commit 71d10a788f6d198e4fc188d6fd2cdc4f285dbd7a)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/time/qdatetime.cpp                      | 2 ++
 tests/auto/corelib/time/qdatetime/tst_qdatetime.cpp | 6 ++++++
 2 files changed, 8 insertions(+)

diff --git a/src/corelib/time/qdatetime.cpp b/src/corelib/time/qdatetime.cpp
index 2efa4b3acf0..d6a827b4966 100644
--- a/src/corelib/time/qdatetime.cpp
+++ b/src/corelib/time/qdatetime.cpp
@@ -200,6 +200,8 @@ static ParsedRfcDateTime rfcDateImpl(QStringView s)
             monthIndex = 0;
             yearIndex = words.size() > 3 && words.at(2).contains(colon) ? 3 : 2;
         }
+        if (words.at(yearIndex).size() != 4)
+            return result;
 
         int dayOfWeek = 0;
         if (!dayName.isEmpty()) {
diff --git a/tests/auto/corelib/time/qdatetime/tst_qdatetime.cpp b/tests/auto/corelib/time/qdatetime/tst_qdatetime.cpp
index 661f15c22d5..76a55ecf965 100644
--- a/tests/auto/corelib/time/qdatetime/tst_qdatetime.cpp
+++ b/tests/auto/corelib/time/qdatetime/tst_qdatetime.cpp
@@ -3066,6 +3066,12 @@ void tst_QDateTime::fromStringDateFormat_data()
         << Qt::RFC2822Date << QDateTime();
     QTest::newRow("RFC 2822 with day date only") << u"Fri, 01 Nov 2002"_s
         << Qt::RFC2822Date << QDateTime();
+    QTest::newRow("RFC 2822 with obsolete 2 digit year")
+            << u"Mon, 07 Feb 22 11:48:12 -0500"_s << Qt::RFC2822Date << QDateTime();
+    QTest::newRow("RFC 2822 with valid year")
+            << u"Mon, 07 Feb 2022 11:48:12 -0500"_s << Qt::RFC2822Date
+            << QDateTime(QDate(2022, 2, 7), QTime(11, 48, 12),
+                         QTimeZone::fromSecondsAheadOfUtc(-5 * 60 * 60));
 
     QTest::newRow("RFC 2822 malformed time (truncated)")
         << u"01 Nov 2002 0:"_s << Qt::RFC2822Date << QDateTime();
-- 
2.51.2

